src/tactic/rename_var.lean

```diff
@@ -0,0 +1,70 @@
+/-
+Copyright (c) 2019 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot
+-/
+
+import tactic.interactive
+
+/-!
+# Rename bound variable tactic
+
+This files defines a tactic `rename_var` whose main purpose is to teach
+renaming of bound variables.
+
+* `rename_var old new` renames variable `old` to `new` in the goal.
+* `rename_var old new at h` does the same in hypothesis `h`.
+
+```lean
+example (P : ℕ →  ℕ → Prop) (h : ∀ n, ∃ m, P n m) : ∀ l, ∃ m, P l m :=
+begin
+  rename_var n q at h, -- h is now ∀ (q : ℕ), ∃ (m : ℕ), P q m,
+  rename_var m n, -- goal is now ∀ (l : ℕ), ∃ (n : ℕ), P k n,
+  exact h -- Lean does not care about those bound variable names
+end
+```lean
+
+## Tags
+
+teaching, tactic
+-/
+open expr
+
+/-- Rename bound variable `old` to `new` in an `expr`-/
+meta def expr.rename_var (old new : name) : expr → expr
+| (pi n bi t b) := (pi (if n = old then new else n) bi (expr.rename_var t) (expr.rename_var b))
```

Comment:
As my comments on the docs indicate: if this is used on the term `∀ x, (P x ∧ ∀ x, Q x)`, it will rename both `x`s. Is that intended? The alternative is stopping as soon as one renaming is done. (Similarly for `(∀ x, P x) ∧ (∀ x, Q x)` -- should it rename one or both?) It makes this function slightly more complicated but is possibly more intuitive.

Reply:
This doesn't come up in real examples so I didn't think about it. I think the current behavior is slightly more intuitive. I wouldn't mind switching to your suggestion since it wouldn't affect my real use cases, but I wouldn't know how to code that more complicated function.

Reply:
Let me be more explicit: I'm not modifying the docstrings until you'll decide what to do with this issue.

Reply:
I think it certainly could come up in real examples in the `(∀ x, P x) ∧ (∀ x, Q x)` case. But then, defaulting to renaming the left `x` is also confusing; we'd need to support renaming a certain occurrence. I think it's not worth the effort.

Avoiding nested ones isn't hard, and the "change only the outermost" story isn't asymmetric. You just need to change the `pi` and `lam` lines to
```suggestion
| (pi n bi t b) := if n = old then pi new bi t b else pi n bi t.rename_var b.rename_var
```
But without the change for non-nested repeated names, the story is still complicated. ("renames all first occurrences of the bound variable ..."? Not super clear.) Up to you if you want to do this.